### PR TITLE
fix: ingress class not effect in resource sync logic

### LINKED ISSUE #1310: bug: ingress filter is not handled in resource sync logic
### Issue description

Apisix ingress controller can filter other ingress controller object by setting ingressclassname. But it seems not work in 1.5.0-rc1. I checked the code and found the filter is implement by 

```
func (c *ingressController) isIngressEffective(ing kube.Ingress) bool {
	var (
		ic  *string
		ica string
	)
	if ing.GroupVersion() == kube.IngressV1 {
		ic = ing.V1().Spec.IngressClassName
		ica = ing.V1().GetAnnotations()[_ingressKey]
	} else if ing.GroupVersion() == kube.IngressV1beta1 {
		ic = ing.V1beta1().Spec.IngressClassName
		ica = ing.V1beta1().GetAnnotations()[_ingressKey]
	} else {
		ic = ing.ExtensionsV1beta1().Spec.IngressClassName
		ica = ing.ExtensionsV1beta1().GetAnnotations()[_ingressKey]
	}

	// kubernetes.io/ingress.class takes the precedence.
	if ica != "" {
		return ica == c.controller.cfg.Kubernetes.IngressClass
	}
	if ic != nil {
		return *ic == c.controller.cfg.Kubernetes.IngressClass
	}
	return false
}
```

and called in onadd, ondelete, onupdate

but in the new func 

```
func (c *ingressController) ResourceSync() {
	objs := c.controller.ingressInformer.GetIndexer().List()
	for _, obj := range objs {
		key, err := cache.MetaNamespaceKeyFunc(obj)
		if err != nil {
			log.Errorw("found Ingress resource with bad meta namespace key", zap.String("error", err.Error()))
			continue
		}
		if !c.controller.isWatchingNamespace(key) {
			continue
		}
		ing := kube.MustNewIngress(obj)
		c.workqueue.Add(&types.Event{
			Type: types.EventAdd,
			Object: kube.IngressEvent{
				Key:          key,
				GroupVersion: ing.GroupVersion(),
			},
		})
	}
}
```

it did't do such check, is this a bug?

### Environment

- your apisix-ingress-controller version (output of apisix-ingress-controller version --long):
- your Kubernetes cluster version (output of kubectl version):
- if you run apisix-ingress-controller in Bare-metal environment, also show your OS version (uname -a):

### PR DESCRIPTION
**Pull request by xiangtianyu**
### Type of change:

- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### What this PR does / why we need it:

fix #1310 
### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

File: pkg/providers/ingress/ingress.go [hunk 1]
```diff
@@ -428,7 +428,20 @@ func (c *ingressController) ResourceSync() {
...
+		if valid {
+			log.Debugw("ingress add event arrived",
+				zap.Any("object", obj),
+			)
+		} else {
+			log.Debugw("ignore noneffective ingress add event",
+				zap.Any("object", obj),
+			)
+			return
+		}
```
Comment (tao12345666333) [inline pkg/providers/ingress/ingress.go:443]: We can simplify the code here, right?

**Review by tao12345666333 — COMMENTED** [review]
thanks for your contribution!

File: pkg/providers/ingress/ingress.go (hunk 1 above)
Comment (xiangtianyu) [reply pkg/providers/ingress/ingress.go:443]: Are you means to summarize these "valid judge" logic to common codes?

Comment (tao12345666333) [reply pkg/providers/ingress/ingress.go:443]: yes, we don't need judge `valid` and print log.

Comment (xiangtianyu) [reply pkg/providers/ingress/ingress.go:443]: > yes, we don't need judge `valid=true` and print log.

Ok, i'll modify it

File: pkg/providers/ingress/ingress.go [hunk 2]
```diff
@@ -286,12 +286,7 @@ func (c *ingressController) onAdd(obj interface{}) {
 	}
 
 	ing := kube.MustNewIngress(obj)
-	valid := c.isIngressEffective(ing)
-	if valid {
-		log.Debugw("ingress add event arrived",
-			zap.Any("object", obj),
-		)
-	} else {
+	if !c.isIngressEffective(ing) {
```
Comment (AlinsRan) [inline pkg/providers/ingress/ingress.go:289]: These places do not need to be changed, and the debug log needs to be kept.

File: pkg/providers/ingress/ingress.go [hunk 3]
```diff
@@ -324,13 +319,7 @@ func (c *ingressController) onUpdate(oldObj, newObj interface{}) {
...
 		return
 	}
-	valid := c.isIngressEffective(curr)
-	if valid {
-		log.Debugw("ingress update event arrived",
-			zap.Any("new object", newObj),
-			zap.Any("old object", oldObj),
-		)
-	} else {
+	if !c.isIngressEffective(curr) {
```
Comment (AlinsRan) [inline pkg/providers/ingress/ingress.go:322]: Ditto.

File: pkg/providers/ingress/ingress.go [hunk 4]
```diff
@@ -428,7 +412,15 @@ func (c *ingressController) ResourceSync() {
 		if !c.namespaceProvider.IsWatchingNamespace(key) {
 			continue
 		}
+
 		ing := kube.MustNewIngress(obj)
+		if !c.isIngressEffective(ing) {
+			log.Debugw("ignore noneffective ingress add event",
```
Comment (AlinsRan) [inline pkg/providers/ingress/ingress.go:418]: Only the log of the sync object should be printed, and the sync should not care about the ingresssclass log.

**Comment by tao12345666333** [comment]
I agree with @AlinsRan , in this PR you can just keep the changes in the ResourceSync function, which will help me cherry-pick it to other versions.

File: pkg/providers/ingress/ingress.go (hunk 4 above)
Comment (xiangtianyu) [reply pkg/providers/ingress/ingress.go:418]: I misunderstanding sonething. And i will modify it soon

**Review by lingsamuel — APPROVED** [review]

File: pkg/providers/ingress/ingress.go [hunk 5]
```diff
@@ -429,6 +429,15 @@ func (c *ingressController) ResourceSync() {
...
 		ing := kube.MustNewIngress(obj)
+		valid := c.isIngressEffective(ing)
+		if valid {
+			log.Debugw("ingress add event arrived",
+				zap.Any("object", obj),
+			)
+		} else {
+			// resource sync do not care about the ingress class mismatch
+			return
+		}
```
Comment (AlinsRan) [inline pkg/providers/ingress/ingress.go:440]:
```suggestion
		if !c.isIngressEffective(ing) {
			return
		}
		log.Debugw("ingress add event arrived",
			zap.Any("object", obj),
		)
```

**Review by AlinsRan — APPROVED** [review]

**Review by tao12345666333 — APPROVED** [review]
LGTM

**Pull request merged**